### PR TITLE
ResolveHostname fixes

### LINKED ISSUE #1179: model.Config ResolveHostname() always appends namespace
Creating the following rule yaml for a consul service with istioctl:
```type: route-rule
name: reviews-default
domain: service.consul
spec:
  destination:
    name: reviews
  precedence: 1
  route:
    - labels:
         version: v1
      weight: 75
    - labels:
          version: v2
      weight: 25
```

Results in the following rule 
```
{
    "type": "route-rule",
    "name": "reviews-default",
    "namespace": "default",
    "domain": "cluster.local",
    "resourceVersion": "71",
    "Spec": {
      "destination": {
        "name": "reviews"
      },
      "precedence": 1,
      "route": [
        {
          "labels": {
            "version": "v1"
          },
          "weight": 75
        },
        {
          "labels": {
            "version": "v2"
          },
          "weight": 25
        }
      ]
    }
  }
```

ResolveHostname() returns `reviews.default.svc.cluster.local` when the desired output should look like `reviews.service.consul`.  Consul, like Eureka, has no concept of namespace, and so it is not a factor in service conversion.

A temporary workaround is to define the rule as follows:
```
type: route-rule
name: reviews-default
spec:
  destination:
    name: reviews
    domain: consul
    namespace: service
  precedence: 1
  route:
    - labels:
         version: v1
      weight: 75
    - labels:
          version: v2
      weight: 25
```

### PR DESCRIPTION
**Pull request by GregHanson**
**What this PR does / why we need it**:
now uses FQDN if set.  one of destination.name or destination.service must be specified for rules.  Resolves rule matching for egress route rules and consul route rules.  `destination.service` has precedence over `destination.name` 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
resolves #1179 

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: model/validation.go [hunk 1]
```diff
@@ -235,9 +235,9 @@ func ValidateFQDN(fqdn string) error {
 
 // ValidateIstioService checks for validity of a service reference
 func ValidateIstioService(svc *proxyconfig.IstioService) (errs error) {
-	if svc.Name == "" {
+	if svc.Name == "" && svc.Service == "" {
 		errs = multierror.Append(errs, errors.New("name is mandatory for a service reference"))
```
Comment (rshriram) [inline model/validation.go]: Change error message: name or service

**Comment by GregHanson** [comment]
works locally for me:
```
=================== Tearing down infrastructure =====================
(default infra)

I0908 15:37:24.434241    9792 kubernetes.go:55] Deleted namespace istio-test-tl0s4
I0908 15:37:24.479721    9792 kubernetes.go:55] Deleted namespace istio-test-f9tmq
I0908 15:37:24.479777    9792 driver.go:169] 

=================== Passed all tests! =====================
tests: [http-reachability http2-reachability tcp-reachability ingress egress routing-rules zipkin], count: 1

.....

=================== Tearing down infrastructure =====================
(auth infra)

I0908 15:39:47.434049    9792 kubernetes.go:55] Deleted namespace istio-test-tqcwk
I0908 15:39:47.481898    9792 kubernetes.go:55] Deleted namespace istio-test-w38lr
I0908 15:39:47.481938    9792 driver.go:169] 

=================== Passed all tests! =====================
tests: [http-reachability http2-reachability tcp-reachability ingress egress routing-rules zipkin], count: 1

I0908 15:39:47.482176    9792 driver.go:169] 

=================== Passed infrastructure tests! =====================
([]main.infra) (len=2 cap=2) {
 (main.infra) {
  Name: (string) (len=15) "(default infra)",
  Hub: (string) (len=8) "gihanson",
  Tag: (string) (len=17) "dev_170908_153443",
  MixerImage: (string) (len=67) "gcr.io/istio-testing/mixer:986c6e1eb9c7dcff42e3df5aefa9ce858ece885e",
  CaImage: (string) (len=70) "gcr.io/istio-testing/istio-ca:2baec6baacecbd516ea0880573b6fc3cd5736739",
  Namespace: (string) "",
  IstioNamespace: (string) "",
  Verbosity: (int) 2,
  apps: (map[string][]string) <nil>,
  Auth: (istio_proxy_v1_config.ProxyMeshConfig_AuthPolicy) NONE,
  Mixer: (bool) true,
  Ingress: (bool) true,
  Egress: (bool) true,
  Zipkin: (bool) true,
  DebugPort: (int) 0,
  checkLogs: (bool) true,
  namespaceCreated: (bool) false,
  istioNamespaceCreated: (bool) false,
  debugImagesAndMode: (bool) true,
  UseInitializer: (bool) false,
  InjectConfig: (*inject.Config)(<nil>),
  UseAdmissionWebhook: (bool) false,
  AdmissionServiceName: (string) (len=20) "istio-pilot-external",
  config: (model.IstioConfigStore) <nil>
 },
 (main.infra) {
  Name: (string) (len=12) "(auth infra)",
  Hub: (string) (len=8) "gihanson",
  Tag: (string) (len=17) "dev_170908_153443",
  MixerImage: (string) (len=67) "gcr.io/istio-testing/mixer:986c6e1eb9c7dcff42e3df5aefa9ce858ece885e",
  CaImage: (string) (len=70) "gcr.io/istio-testing/istio-ca:2baec6baacecbd516ea0880573b6fc3cd5736739",
  Namespace: (string) "",
  IstioNamespace: (string) "",
  Verbosity: (int) 2,
  apps: (map[string][]string) <nil>,
  Auth: (istio_proxy_v1_config.ProxyMeshConfig_AuthPolicy) MUTUAL_TLS,
  Mixer: (bool) true,
  Ingress: (bool) true,
  Egress: (bool) true,
  Zipkin: (bool) true,
  DebugPort: (int) 0,
  checkLogs: (bool) true,
  namespaceCreated: (bool) false,
  istioNamespaceCreated: (bool) false,
  debugImagesAndMode: (bool) true,
  UseInitializer: (bool) false,
  InjectConfig: (*inject.Config)(<nil>),
  UseAdmissionWebhook: (bool) false,
  AdmissionServiceName: (string) (len=20) "istio-pilot-external",
  config: (model.IstioConfigStore) <nil>
 }
}
```

File: model/validation.go [hunk 2]
```diff
@@ -235,9 +235,9 @@ func ValidateFQDN(fqdn string) error {
 
 // ValidateIstioService checks for validity of a service reference
 func ValidateIstioService(svc *proxyconfig.IstioService) (errs error) {
-	if svc.Name == "" {
-		errs = multierror.Append(errs, errors.New("name is mandatory for a service reference"))
-	} else if !IsDNS1123Label(svc.Name) {
+	if svc.Name == "" && svc.Service == "" {
```
Comment (kyessenov) [inline model/validation.go]: check that if service is specified then name, namespace, domain are not.
check that service is FQDN/wildcard mask (see egress rule validation)

**Comment by rshriram** [comment]
Filed bug to add validation. Greg has left for the day. @vadimeisenbergibm will pick up more of the egress routing issue on Sunday.

**Pull request merged**